### PR TITLE
Improve support for LLD_REPORT_UNDEFINED

### PR DESCRIPTION
**Pull request by sbc100**
Mark EM_JS function with import attributes so that wasm-ld will allow
them to be undefined.

### CODE REVIEW

**Comment by sbc100** [comment]
OK, this change is finally good to go and rebased onto master.

**Comment by sbc100** [comment]
Because I split it up into lots of smaller PRs this one ended up being petty tiny.

File: emcc.py [hunk 1]
```diff
@@ -1214,6 +1217,11 @@ def check(input_file):
     if shared.Settings.ASSERTIONS:
       shared.Settings.STACK_OVERFLOW_CHECK = 2
 
+    if shared.Settings.LLD_REPORT_UNDEFINED:
+      # Reporting undefined symbols at wasm-ld time requires us to know if we have `main` function
```
Comment (kripken) [inline emcc.py:1221]: `the` after `have`

File: src/settings_internal.js [hunk 2]
```diff
@@ -188,3 +188,7 @@ var SEPARATE_DWARF = 0;
 
 // New WebAssembly exception handling (experimental)
 var EXCEPTION_HANDLING = 0;
+
+// Used internally when running the JS compiler simply to generate list of all
+// JS symbols.
+var CALCULATE_ONLY = 0;
```
Comment (kripken) [inline src/settings_internal.js:194]: this name isn't descriptive enough I feel? It doesn't say what is calculated. How about `ONLY_CALC_JS_SYMBOLS`?

File: system/include/emscripten/em_js.h [hunk 3]
```diff
@@ -56,7 +56,7 @@
 
 #define EM_JS(ret, name, params, ...)          \
   _EM_JS_CPP_BEGIN                             \
-  extern ret name params;                      \
+  extern ret name params EM_IMPORT(name);      \
```
Comment (kripken) [inline system/include/emscripten/em_js.h:59]: what is this change?

File: tests/runner.py [hunk 4]
```diff
@@ -426,11 +426,17 @@ def setUpClass(cls):
...
+    self.emcc_args = []
     self.save_dir = EMTEST_SAVE_DIR
     self.env = {}
     self.temp_files_before_run = []
 
+    # Ideally all our tests code would be built with -Wall, -Werror and -s STRICT.  For now
+    # we just enable a subset of the options that STRICT would enable.
+    self.emcc_args.append('-Werror')
+    self.set_setting('LLD_REPORT_UNDEFINED', 1)
+    self.set_setting('IGNORE_MISSING_MAIN', 0)
```
Comment (kripken) [inline tests/runner.py:438]: do we have enough tests that check the default values of these flags (which is what our users use mostly)? I guess we have `other`, maybe that's enough.

File: tests/test_core.py [hunk 5]
```diff
@@ -8816,10 +8817,13 @@ def test_undefined_main(self):
     # building library code that has no main.
     # TODO(sbc): Simplify the code by making this an opt-in feature.
     # https://github.com/emscripten-core/emscripten/issues/9640
+
     src = '''
     #include <emscripten.h>
     EMSCRIPTEN_KEEPALIVE void foo() {}
     '''
+    # Clear out the default tests settings since they include IGNORE_MISSING_MAIN=0
+    self.settings_mods = {}
```
Comment (kripken) [inline tests/test_core.py:8826]: I've never seen us do this in a test before, it seems wrong to me. That's the internal state, isn't it? Why not modify the state using `set_setting`?

File: tests/runner.py (hunk 4 above)
Comment (sbc100) [reply tests/runner.py:438]: In general for each setting we have a test that test the effect of it both enabled and disabled.  So I think we should be fine.  But you raise a good point.  I think maybe this is the wrong approach.   Maybe we should consider this one of the "modes" in which we run the test suite.   I'll revert his part of the change for now.

(BTW, I would like to see all tests (including other) use this system and honor these default flags uses by the test suite)

File: tests/test_core.py (hunk 5 above)
Comment (sbc100) [reply tests/test_core.py:8826]: Revert all this stuff.

File: tests/runner.py (hunk 4 above)
Comment (kripken) [reply tests/runner.py:438]: Another option is to move this test to `other` if you want to control all the commands 100% manually. `core` is useful for things that you trust the default core modes to cover, as I see it.

File: src/settings_internal.js [hunk 6]
```diff
@@ -188,3 +188,8 @@ var SEPARATE_DWARF = 0;
 
 // New WebAssembly exception handling (experimental)
 var EXCEPTION_HANDLING = 0;
+
+// Used internally when running the JS compiler simply to generate list of all
+// JS symbols. This is used by LLD_REPORT_UNDEFINED to generate a list of all
+// JS library symbols.
+var ONLY_CALCUALTE_JS_SYMBOLS = 0;
```
Comment (kripken) [inline src/settings_internal.js:195]: typo in `CALCULATE`

File: tools/shared.py [hunk 7]
```diff
@@ -1813,8 +1817,7 @@ def link_lld(args, target, opts=[], all_external_symbols=None):
       ]
       use_start_function = Settings.STANDALONE_WASM
 
-      if not use_start_function:
-        expect_main = '_main' in Settings.EXPORTED_FUNCTIONS
+      if not use_start_function and '--relocatable' not in opts and '-r' not in opts:
```
Comment (kripken) [inline tools/shared.py:1820]: Why is relocation relevant here?

**Comment by sbc100** [comment]
This change is now much smaller. PTAL.

File: emcc.py [hunk 8]
```diff
@@ -543,13 +540,15 @@ def get_all_js_library_funcs(temp_files):
...
-        # such as __wasi_fd_write which are defined with import_name attibutes
-        # but it currently does.  Remove this once we fix wasm-ld.
-        library_fns_list.append('__wasi_' + name)
   finally:
+    shared.Settings.ONLY_CALC_JS_SYMBOLS = False
     shared.Settings.INCLUDE_FULL_LIBRARY = old_full
-    shared.Settings.LINKABLE = old_linkable
+
+  # Some of these functions are actually not JS functions at all but fallbacks
+  # missing undefined native symbols.  Ignore these.
```
Comment (kripken) [inline emcc.py:548]: `fallbacks [to]`?

But why are these special? Is this just fastcomp?

File: system/include/emscripten/em_js.h (hunk 3 above)
Comment (kripken) [reply system/include/emscripten/em_js.h:59]: Still curious about this?

File: tests/test_core.py [hunk 9]
```diff
@@ -1624,6 +1624,7 @@ def test_iostream_ctors(self):
 
   @also_with_standalone_wasm
   def test_ctors_no_main(self):
+    self.emcc_args.append('--no-entry')
```
Comment (kripken) [inline tests/test_core.py:1627]: why is this needed? will users need to do it too?

File: emcc.py (hunk 8 above)
Comment (sbc100) [reply emcc.py:548]: I think the problem is that sometimes these sybmols were coming from native code and sometimes from JS.  I'll try to revert this part for now as it was only causes a few tests to fail IIRC.

File: system/include/emscripten/em_js.h (hunk 3 above)
Comment (sbc100) [reply system/include/emscripten/em_js.h:59]: So this macro is used for defining em_js functions.   At link time these are always expected to be undefined since they are implemented in JS.   Unlike the JS library functions  we can't calculate the list of all known EM_JS symbols ahead of time so we use this method of marking them as external.

In the long run this method is preferable, and ideally all JS library functions would be marked like this in the source code.  However that is much larger and more ambitious change.

File: tests/test_core.py (hunk 9 above)
Comment (sbc100) [reply tests/test_core.py:1627]: No, this is only needs to that the test still passed in STRICT mode.

Comment (sbc100) [reply tests/test_core.py:1627]: Eventually thought, this is the plan, that users would either need to pass an empty list of EXPORTED_FUNCTIONS of --no-entry (one or the other).   We agreed on this already I believe.

Comment (kripken) [reply tests/test_core.py:1627]: This test doesn't pass `STRICT`, though?

Comment (kripken) [reply tests/test_core.py:1627]: I'm asking, is this for future use, or was it needed now in a way that I'm missing?

Comment (sbc100) [reply tests/test_core.py:1627]: I've added a new mode below called `strict` that runs that entire suite in strict mode.  Like safe heap mode it doesn't run in CI but is useful for keeping out test suite clean.

Comment (sbc100) [reply tests/test_core.py:1627]: See below.

Comment (kripken) [reply tests/test_core.py:1627]: Oh, if it's just for strict mode, how about an if on that? But if we want to do this by default later I guess that's pointless, so sgtm as is.

File: tests/test_core.py [hunk 10]
```diff
@@ -6951,8 +6952,8 @@ def test_demangle_stacks(self, extra_args):
     self.emcc_args += extra_args
     self.set_setting('DEMANGLE_SUPPORT', 1)
     self.set_setting('ASSERTIONS', 1)
-    # when optimizing function names are not preserved by default.
-    if '-O' in str(self.emcc_args):
+    # unless `-g` is pass function names are not preserved by default.
```
Comment (kripken) [inline tests/test_core.py:6955]: `is pass` => `is passed`

but i think the comment is misleading after the change. as the code shows, `--profiling-funcs` is enough, so `-g` is not necessary. how about just `ensure function names are preserved`?
